### ISSUE TITLE
Can we compare two pdf files using PyMuPDF (visually and text)

### ISSUE THREAD

**Issue by kvrameshreddy** [pymupdf/PyMuPDF#673]
Hi,

I have extracted text from two pdf files separately and then comparing for the changes in text  using "difflib" and annotating the variations using "page.searchFor( )". The issue in this case is, if the file has same text at multiple positions i am getting annotations at all the places ( even the difference is not there in the pdfs).

Can you help me in this? Please suggest me any work around for the above use case.

thankyou.

i am using latest version of PyMuPDF on pyhton 3.7 with windows.

**Comment by JorjMcKie** [comment]
What data are you comparing when using difflib? The PDF files themselves, or some type of extracted text?
The principle problem here is that you can have seemingly equal PDF which will never compare equal, because you can position each character separately using some arbitrary sequence.
Like PDF 1 has string "Jorj McKie" and PDF 2 has "e", "K", "J", "c" "i", "M", "j", "o", "r", but each of the single letters sitting at the exact same positions. I've seen this happening, although it's rare of course. Text extractions will return the text in sequence as encountered ... and you would need to sort (and potentially then re-synthesize the words) before comparing.

As a practical approach (not assuming the worst scenario) you could take output of ``page.getText("blocks")`` or ``page.getText("words")`` and compare them. If differences occur you then already have the coordinates for annotation.

`Page.searchFor` has its peculiarities: if there is text like "abc", "abcabc" and you search for "abc", you will see only **two** rectangles, one for "abc" and one for "abcabc". And more like that.

**Comment by kvrameshreddy** [comment]
Thankyou @JorjMcKie. I will work on that.

**Comment by JorjMcKie** [comment]
Please come back with questions or how you resolved it.
I would like to include your experience in the recipes.
A similar question some time ago resulted in "visually" comparing the pages' pixmaps and encircling the deviating pixels ...

**Comment by kvrameshreddy** [comment]
Definitely. In general visual comparison can be done using opencv and SSIM which does pixel by pixel. In my case i am having two PDF files which contain graphs, tables and text. I need to compare for text variations and image variations and mark the differences. Even the position of text changes in the pdf i should not mark it as different.

**Comment by JorjMcKie** [comment]
I see.
You may be interested in the new utility module currently contained in https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/extract-graphics.
It reads out the graphics on a page (drawn lines, rectangles, curves ...) and returns those data (so-called "paths" in PDF terminology) organized in Python dicts.

BTW I am considering to include this module in standard PyMuPDF once it has stabilized a bit more.

**Comment by kvrameshreddy** [comment]
Hi @JorjMcKie  Can you help me in this. 
I have few pdfs in which i am unable to extract images out. i have used "doc.getPageImageList()" which gives empty list.
When i try to extract json of the pdfs, json does not contains image references.  

below is the code i have tried and attached the respective pdf files.


doc = fitz.open(pdffile)
doc.getPageImageList(0)
#doc.metadata
#doc.getPageImageList(0)
for i in range(len(doc)):
    for img in doc.getPageImageList(i):
        xref=img[0]
        print(xref)
        pix=fitz.Pixmap(doc,xref)
        if pix.n < 5:
            pix.writePNG("pp%s-%s.png" % (i,xref))
        else:
            pix1=fitz.Pixmap(fitz.csRGB, pix)
            pix1.writePNG("p%s-%s.png" % (i,xref))
            pix1 = None
        pix = None

[Customers_pre.pdf](https://github.com/pymupdf/PyMuPDF/files/5312886/Customers_pre.pdf)
[Trellis1_Pre.pdf](https://github.com/pymupdf/PyMuPDF/files/5312889/Trellis1_Pre.pdf)

**Comment by JorjMcKie** [comment]
File ``Customers_pre.pdf`` does have one image on page 0. It can be extracted, too:
```python
>>> page.getImageList(True)
[(6, 5, 1600, 620, 8, 'DeviceRGB', '', 'FXX1', 'DCTDecode', 0)]
```
Re-constructing the original image is a bit complicated in this case, because it requires takng care of the transparency mask in xref 5:
```python
>>> from PIL import Image
>>> import io
>>> img6 = io.BytesIO(doc.extractImage(6)["image"])  # read image binary via io.BytesIO
>>> img5 = io.BytesIO(doc.extractImage(5)["image"])
>>> # the final PIL image wil be this:
>>> img = Image.new("RGBA", (1600, 620))
>>> pimg6 = Image.open(img6)  # open image binary as PIL image
>>> pimg5 = Image.open(img5)
>>> img.paste(pimg6, None, pimg5)  # apply the transparency
>>> # now save image as required:
>>> img.save("image6.png")
```

**Comment by kvrameshreddy** [comment]
Thankyou very much @JorjMcKie for taking time to clear my issue.

**Comment by kvrameshreddy** [comment]
Hi @JorjMcKie ,
When i use  extractGraphic i am getting the output in inverse. Please help me in this.

I have attached the input file and output file for your reference.

[Customers1_post.pdf](https://github.com/pymupdf/PyMuPDF/files/5327111/Customers1_post.pdf)
[Customers1_Post-new.pdf](https://github.com/pymupdf/PyMuPDF/files/5327112/Customers1_Post-new.pdf)

Thankyou.

**Comment by JorjMcKie** [comment]
Pleased to see you are using it, so ther is a real interest ... 😎
However my stuff is still buggy. It does does recognize everything and in addition in your case it swallowed a matrix which turns the page upside down. That's the reason for your result.

In the meantime I have developed a completely different approach, which is a lot less buggy. It produces this result:
[Customers1_post-new2.pdf](https://github.com/pymupdf/PyMuPDF/files/5327611/Customers1_post-new2.pdf)

I also have integrated the functionality in PyMuPDF. To retrieve the list of drawing paths, you just need to do this:
```python
paths = page.getDrawings()
```
It also works for XPS, EPUB, etc. not just PDF as it was the case earlier. I can send you a wheel for your platform if you are interested.

**Comment by kvrameshreddy** [comment]
This looks very nice. Can you please explain me how to get Drawing out of a pdf. Please send me the wheel file 
i am working on windows with python 3.7.3

**Comment by JorjMcKie** [comment]
> Can you please explain me how to get Drawing out of a pdf.

This would work exactly like before, except that you do not need to import anything extra. The function has become a page method.

> Please send me the wheel file i am working on windows with python 3.7.3

Are you using Python 64bit?

**Comment by kvrameshreddy** [comment]
yes @JorjMcKie  python 64bit

**Comment by JorjMcKie** [comment]
[PyMuPDF-1.17.8-cp37-cp37m-win_amd64.zip](https://github.com/pymupdf/PyMuPDF/files/5327856/PyMuPDF-1.17.8-cp37-cp37m-win_amd64.zip)

To install this pre-version, do this:
1. Rename the file extension ".zip" back to ".whl" - needed to do this because of a GitHub restriction.
2. Then execute ``python -m pip install -U PyMuPDF-1.17.8-cp37-cp37m-win_amd64.whl``.

There also is a minor change to the output test script:
![grafik](https://user-images.githubusercontent.com/8290722/95089561-48503380-06f2-11eb-8cd6-6067901a7b3e.png)

The lineCap parameter is now a 3-tuple - was an int before. This obviously is for document types other than PDF, so I just take the maximum value for the time being.

**Comment by kvrameshreddy** [comment]
Thankyou very much. I will work on this.

**Comment by JorjMcKie** [comment]
As a general note:

* Although much better, this is still under development and should be regarded as _experimental_.
* All rectangles / quadrilaterals are coded as a series of lines. I am still working on how to recover from this.
* If a figure has a border and is filled (like circles or rectangles), **_two_** separate paths will be created: one for drawing the border (stroke) and a second one for filling. This is another thing I will try to join again.
* Like before: opacity and blend mode cannot be reproduced. It would be a major effort to reproduce these. So everything will be drawn with no transparency.
* A document page can have been crafted in a very complex manner. Drawings may be set to being (partly) visible, grouped together in some way, etc. I am ignoring all this and just report all drawings I am encountering. So the output PDF page generated will usually show more than what was visible in the original: you file is an example for this.
* When looking at your example, it seems the line width is too thick: I may be able to correct this in some way.

**Comment by kvrameshreddy** [comment]
ya I also observed the line thickness. Waiting for the full version. Thankyou @JorjMcKie

**Comment by JorjMcKie** [comment]
This issue is fixed with v1.18.0.
